### ISSUE TITLE
Cannot load MIB file

### ISSUE THREAD

**Issue by zhum** [swisscom/ruby-netsnmp#48]
Hello! I try to use netsnmp gem with Panasas hardware. Here is a MIB-file, which cannot be parsed (but `snmpget/snmpwalk` work fine with it), I get this error:

```
...
        5: from /home/serg/.gem/ruby/gems/netsnmp-0.4.0/lib/netsnmp/mib.rb:82:in `load'
        4: from /home/serg/.gem/ruby/gems/netsnmp-0.4.0/lib/netsnmp/mib.rb:97:in `do_load'
        3: from /home/serg/.gem/ruby/gems/netsnmp-0.4.0/lib/netsnmp/mib.rb:97:in `synchronize'
        2: from /home/serg/.gem/ruby/gems/netsnmp-0.4.0/lib/netsnmp/mib.rb:97:in `block in do_load'
        1: from /home/serg/.gem/ruby/gems/parslet-2.0.0/lib/parslet/atoms/base.rb:49:in `parse'
/home/serg/.gem/ruby/gems/parslet-2.0.0/lib/parslet/cause.rb:70:in `raise': Failed to match sequence (SPACE{0, } MODULES?) at line 2 char 1. (Parslet::ParseFailed)
```
It comes with several other MIB-files and I load them in recommended order (there is special README file), I can send them also if needed. But I think there is something unusual in the file, but parser output cannot tell me where...

[PANASAS-HW-MIB-V1.mib.txt](https://github.com/swisscom/ruby-netsnmp/files/6523571/PANASAS-HW-MIB-V1.mib.txt)

**Comment by HoneyryderChuck** [comment]
Hi, and thank you for the report.

The MIB parser is a recent addition, and might need to be adjusted for patterns is not yet able to parse. I managed to make it parse the default set, but there's certainly other custom ones that might fall through the cracks.

I'm not able to prioritize this fix right now, So if you need this, I kindly ask you to do it yourself. This what you can do:

1. copy the MIB file to a location where you can edit it.
2. set the simple ruby script:

```ruby
require "netsnmp"
require "netsnmp/mib/parser"

puts NETSNMP::MIB::Parser.new.parse(File.read("path/to/file"))
```
3. eliminate rules from the file until the script above parses.
4. add the necessary the parslet rules

**Comment by zhum** [comment]
I tried to do this, but I get:

```
Traceback (most recent call last):
        4: from mib-test.rb:18:in `<main>'
        3: from mib-test.rb:18:in `each'
        2: from mib-test.rb:21:in `block in <main>'
        1: from /home/serg/.gem/ruby/gems/parslet-2.0.0/lib/parslet/atoms/base.rb:49:in `parse'
/home/serg/.gem/ruby/gems/parslet-2.0.0/lib/parslet/cause.rb:70:in `raise': Failed to match sequence (SPACE{0, } MODULES?) at line 1 char 1. (Parslet::ParseFailed)
```

So it fails on the first line beginning:
```
PANASAS-HW-MIB-V1 DEFINITIONS ::= BEGIN
```

This line seems to be ok, here is the hexdump:

```
00000000  50 41 4e 41 53 41 53 2d  48 57 2d 4d 49 42 2d 56  |PANASAS-HW-MIB-V|
00000010  31 20 44 45 46 49 4e 49  54 49 4f 4e 53 20 3a 3a  |1 DEFINITIONS ::|
00000020  3d 20 42 45 47 49 4e 0a  0a 2d 2d 09 54 68 69 73  |= BEGIN..--.This|
```

The corresponding rule seems to be not required to be fixed, and I don't know where to dig further...

**Comment by HoneyryderChuck** [comment]
```ruby
require "netsnmp"
mib = <<-OUT
PANASAS-HW-MIB-V1 DEFINITIONS ::= BEGIN
END
OUT
NETSNMP::MIB::Parser.new.parse(mib)
#=> {:name=>"PANASAS-HW-MIB-V1"@0, :exports=>nil, :imports=>nil, :declarations=>nil}
```

It does work. I suggest you add statements inside this block until you find the buggy rule.

**Comment by zhum** [comment]
Found an error - in one SEQUENCE a comma was missed. But snmpget/snmpwalk ignore this.

Now I can import all MIBS, but I cannot resolve oid:
```
NETSNMP::MIB.oid("PANASAS-HW-MIB-V1::panHwNodeOsdStorageServerName")
```
gets nil. While this works:
```
snmpwalk -v1 -c public 10.0.251.1 PANASAS-HW-MIB-V1::panHwNodeOsdStorageServerName
PANASAS-HW-MIB-V1::panHwNodeOsdStorageServerName."2fc601cb100037001" = STRING: "Shelf-1,4"
...
```

How can I debug why resolve does not work?

**Comment by HoneyryderChuck** [comment]
Does `"PANASAS-HW-MIB-V1::panHwNodeOsdStorageServerName"` appear in the `declarations` list, when you load the mib file? Picking up on my example from above:

```ruby
#=> {:name=>"PANASAS-HW-MIB-V1"@0, :exports=>nil, :imports=>nil, :declarations=>nil} # it should appear in the returned structure
```

**Comment by zhum** [comment]
Yes:
```
>ruby ./mib-test.rb | grep panHwNodeOsdStorageServerName
   {:name=>"panHwNodeOsdStorageServerName"@32813,
```

Full declaration:
```
   {:name=>"panHwNodeOsdStorageServerName"@32813,
    :type=>"OBJECT-TYPE"@32844,
    :syntax=>{:value=>"DisplayString"@32868},
    :value=>"panHwNodeEntry 24"@33071},
```

**Comment by zhum** [comment]
"PANASAS-HW-MIB-V1::panHwNodeOsdStorageServerName" - does not resolve.
"panHwNodeOsdStorageServerName" - do resolve... 

How to make the first one to be resolved too?

**Comment by HoneyryderChuck** [comment]
Yes, I was going to suggest you that. However, the previous one should have worked.

I can have a look. Can you post the update mibfile + any changes you might have done to the parser?

**Comment by zhum** [comment]
[PANASAS-HW-MIB-V1.mib.txt](https://github.com/swisscom/ruby-netsnmp/files/6560360/PANASAS-HW-MIB-V1.mib.txt)

Here it is.

**Comment by HoneyryderChuck** [comment]
just released `v0.4.1`
